### PR TITLE
Move ledger write to its own stage

### PR DESCRIPTION
**Pull request by pgarg66**
- Also, rename write_stage to leader_vote_stage, as write functionality
  is moved to a different stage

### CODE REVIEW

**Comment by pgarg66** [comment]
https://github.com/solana-labs/solana/issues/1508

File: src/ledger_write_stage.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
+//! The `ledger_write_stage` module implements the ledger write stage. It
+//! writes entries to the given writer, which is typically a file or
+//! stdout
```
Comment (rob-solana) [inline src/ledger_write_stage.rs]: never stdout anymore

File: src/ledger_write_stage.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                received_entries = n;
+            } else {
+                break;
+            }
+        }
+
+        if let Some(ledger_writer) = ledger_writer {
+            for entries in ventries {
+                for e in &entries {
+                    ledger_writer.write_entry_noflush(&e)?;
```
Comment (rob-solana) [inline src/ledger_write_stage.rs]: I really don't want to share details of flush() with ledger_writer clients, just remove flush() from ledger

**Comment by garious** [comment]
I approve if @rob-solana approves.

File: src/ledger.rs [hunk 3]
```diff
@@ -399,15 +399,9 @@ impl LedgerWriter {
...
 
-    pub fn flush(&mut self) -> io::Result<()> {
-        self.index.flush()?;
-        self.data.flush()?;
-        Ok(())
-    }
-
     pub fn write_entry(&mut self, entry: &Entry) -> io::Result<()> {
         self.write_entry_noflush(&entry)?;
-        self.flush()
```
Comment (rob-solana) [inline src/ledger.rs]: this flush should stay in order to honor the ledger contract?

File: src/ledger_write_stage.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+            if let Ok(n) = entry_receiver.try_recv() {
+                received_entries = n;
+            } else {
+                break;
+            }
+        }
+
+        if let Some(ledger_writer) = ledger_writer {
+            for entries in ventries {
```
Comment (rob-solana) [inline src/ledger_write_stage.rs]: flat_map() to write_entries() instead?

File: src/ledger_write_stage.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                for e in &entries {
+                    ledger_writer.write_entry(&e)?;
+                }
+                if let Some(forwarder) = forwarder {
```
Comment (rob-solana) [inline src/ledger_write_stage.rs]: probably don't want to predicate whether to forward on whether there's a ledger

File: src/ledger.rs [hunk 6]
```diff
@@ -399,15 +399,9 @@ impl LedgerWriter {
         Ok(())
     }
 
-    pub fn flush(&mut self) -> io::Result<()> {
```
Comment (rob-solana) [inline src/ledger.rs]: glad this is gone, but I want to preserve the "write_entry()" or "write_entries()" implies flush

File: src/ledger_write_stage.rs [hunk 7]
```diff
@@ -36,13 +36,12 @@ impl LedgerWriteStage {
         }
 
         if let Some(ledger_writer) = ledger_writer {
+            ledger_writer.write_entries(ventries.iter().flatten())?;
```
Comment (rob-solana) [inline src/ledger_write_stage.rs]: this is rocks!

**Pull request merged**